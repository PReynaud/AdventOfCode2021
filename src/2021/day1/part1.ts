import { isIncreasing } from './common';

export function countNumberOfIncrease(values: number[]): number {
  let result = 0;

  values.forEach((element, index) => {
    if (index > 0) {
      if (isIncreasing(values[index - 1], values[index])) {
        result++;
      }
    }
  });

  return result;
}
