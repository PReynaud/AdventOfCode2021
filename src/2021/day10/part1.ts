export function simplifyLine(value: string): string {
  let currentValue = value;
  let previousValue = currentValue;
  do {
    previousValue = currentValue;
    currentValue = currentValue
      .replace('()', '')
      .replace('[]', '')
      .replace('{}', '')
      .replace('<>', '');
  } while (currentValue && currentValue.length < previousValue.length);
  return currentValue;
}

const openingChars = ['[', '{', '(', '<'];

export function isIncomplete(line: string): boolean {
  return line.split('').every((char) => openingChars.includes(char));
}

function findFirstWrongChar(line: string): string {
  for (let i = 0; i < line.length; i++) {
    if (!openingChars.includes(line[i])) {
      return line[i];
    }
  }
  return '';
}

export function simplifyAndFind(line: string): string {
  return findFirstWrongChar(simplifyLine(line));
}

export function selectIncorrectLines(allLines: string[]): string[] {
  return allLines.filter((oneLine) => !isIncomplete(simplifyLine(oneLine)));
}

function returnValueForChar(char: string): number {
  if (char === ')') return 3;
  if (char === ']') return 57;
  if (char === '}') return 1197;
  if (char === '>') return 25137;
  return 0;
}

export function countIncorrectLines(allLines: string[]): number {
  return selectIncorrectLines(allLines)
    .map(simplifyAndFind)
    .map(returnValueForChar)
    .reduce((acc, value) => acc + value, 0);
}
