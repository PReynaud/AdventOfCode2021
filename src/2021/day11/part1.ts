interface Case {
  value: number;
  hasFlashed: boolean;
}

export type Grid = Case[][];

export function transformRawValuesToGrid(rawValues: string[]): Grid {
  return rawValues.map((oneRow) =>
    oneRow.split('').map((oneValue) => ({
      value: Number(oneValue),
      hasFlashed: false,
    }))
  );
}

interface Position {
  x: number;
  y: number;
}

export function findNeighbours(
  currentPosition: Position,
  xMax: number,
  yMax: number
): Position[] {
  const x = currentPosition.x;
  const y = currentPosition.y;

  return [
    { x: x - 1, y: y - 1 },
    { x: x - 1, y },
    { x: x - 1, y: y + 1 },
    { x, y: y - 1 },
    { x, y: y + 1 },
    { x: x + 1, y: y - 1 },
    { x: x + 1, y },
    { x: x + 1, y: y + 1 },
  ].filter(
    (onePosition) =>
      onePosition.x >= 0 &&
      onePosition.y >= 0 &&
      onePosition.x < xMax &&
      onePosition.y < yMax
  );
}

function getCaseForPosition(grid: Grid, position: Position): Case {
  return grid[position.y][position.x];
}

function increaseValue(grid: Grid, currentPosition: Position): void {
  const currentCase = getCaseForPosition(grid, currentPosition);
  if (currentCase.value < 9 && !currentCase.hasFlashed) {
    currentCase.value++;
  } else if (currentCase.value === 9 && !currentCase.hasFlashed) {
    currentCase.value = 0;
    currentCase.hasFlashed = true;
    increaseNeighbours(grid, currentPosition);
  }
}

export function increaseNeighbours(
  grid: Grid,
  currentPosition: Position
): void {
  const neighbours = findNeighbours(
    currentPosition,
    grid[0].length,
    grid.length
  );
  neighbours.forEach((neighbour) => {
    increaseValue(grid, neighbour);
  });
}

function resetAndCount(grid: Grid): number {
  let result = 0;
  grid
    .flat()
    .flat()
    .forEach((oneCase) => {
      if (oneCase.value === 0) {
        oneCase.hasFlashed = false;
        result++;
      }
    });

  return result;
}

function displayGrid(grid: Grid): void {
  let result = '';
  for (let y = 0; y < grid.length; y++) {
    result += grid[y].map((oneValue) => oneValue.value).join('') + '\n';
  }
  console.log(result);
}

export function runOneStep(grid: Grid): number {
  for (let x = 0; x < grid[0].length; x++) {
    for (let y = 0; y < grid.length; y++) {
      increaseValue(grid, { x, y });
    }
  }

  return resetAndCount(grid);
}

export function runSteps(rawValues: string[], numberOfSteps: number): number {
  const grid = transformRawValuesToGrid(rawValues);
  let numberOfFlashes = 0;
  for (let count = 0; count < numberOfSteps; count++) {
    numberOfFlashes += runOneStep(grid);
  }
  return numberOfFlashes;
}
