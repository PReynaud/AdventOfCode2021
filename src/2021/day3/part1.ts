import { convertBinaryToNumber } from './common';

export function extractGammaRate(values: string[]): string {
  if (values[0]) {
    const maxIndex = values[0].length;
    let result = '';
    for (let i = 0; i < maxIndex; i++) {
      let zeroValues = 0;
      let oneValues = 0;

      values.forEach((oneString) => {
        if (oneString[i] === '0') {
          zeroValues++;
        } else if (oneString[i] === '1') {
          oneValues++;
        }
      });

      result += zeroValues > oneValues ? '0' : '1';
    }
    return result;
  }
  return '';
}

export function invertBites(binaryString: string): string {
  let result = '';
  for (let i = 0; i < binaryString.length; i++) {
    result += binaryString[i] === '0' ? '1' : '0';
  }
  return result;
}

export function extractRates(values: string[]): [string, string] {
  const gammaRate = extractGammaRate(values);
  const epsilonRate = invertBites(gammaRate);
  return [gammaRate, epsilonRate];
}

export function calculatePowerConsumption(values: string[]): number {
  const [gammaRate, epsilonRate] = extractRates(values);
  return convertBinaryToNumber(gammaRate) * convertBinaryToNumber(epsilonRate);
}
