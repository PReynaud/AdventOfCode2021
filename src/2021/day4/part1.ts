import {
  BoardCase,
  calculateScoreWinningBoard,
  checkBoardIsWinner,
  extractBoards,
  extractDrawnNumbers,
  markValuesFromDrawnNumber,
} from './common';

function run(boards: BoardCase[][][], initialNumbers: number[]): number | null {
  let hasWinning = false;
  let numbersToDraw = initialNumbers.slice();
  let currentDrawnNumber: number | null = null;

  while (!hasWinning && numbersToDraw.length > 0) {
    currentDrawnNumber = numbersToDraw[0];
    boards.forEach((board) =>
      markValuesFromDrawnNumber(board, currentDrawnNumber!)
    );
    hasWinning = boards.some((board) => checkBoardIsWinner(board));
    numbersToDraw = numbersToDraw.slice(1);
  }

  if (hasWinning) {
    const winningBoard = boards.find((board) => checkBoardIsWinner(board));
    const score = calculateScoreWinningBoard(winningBoard!);
    return score * currentDrawnNumber!;
  }

  return null;
}

export function extractAndRun(rawValues: string[]): number | null {
  const drawnNumbers = extractDrawnNumbers(rawValues);
  const boards = extractBoards(rawValues);

  return run(boards, drawnNumbers);
}
