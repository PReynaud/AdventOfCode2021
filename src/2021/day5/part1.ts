import {
  calculateOverlapping,
  createEmptyGrid,
  drawStraightLine,
  extractLineInformations,
  Grid,
  isStraightLine,
  Line,
} from './common';

export function drawAllLines(grid: Grid, lines: Line[]): void {
  lines
    .filter((oneLine) => isStraightLine(oneLine))
    .forEach((oneLine) => {
      drawStraightLine(grid, oneLine);
    });
}

export function calculateFromRawValues(rawValues: string[]): number {
  const extracted = extractLineInformations(rawValues);
  const grid = createEmptyGrid(extracted);
  drawAllLines(grid, extracted);
  return calculateOverlapping(grid);
}
