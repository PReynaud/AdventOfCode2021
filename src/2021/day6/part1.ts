import { extractValuesFromRaw } from './common';

export function decreaseValuesAndAddNewFishes(values: number[]): number[] {
  let numberOfNewFishes = 0;
  const newValues = values.map((oneValue) => {
    if (oneValue === 0) {
      numberOfNewFishes++;
      return 6;
    }
    return oneValue - 1;
  });
  return [...newValues, ...Array(numberOfNewFishes).fill(8)];
}

export function calculateNumberOfFish(
  rawValues: string,
  numberOfDays: number
): number {
  let values = extractValuesFromRaw(rawValues);
  for (let i = 0; i < numberOfDays; i++) {
    values = decreaseValuesAndAddNewFishes(values);
  }
  return values.length;
}
