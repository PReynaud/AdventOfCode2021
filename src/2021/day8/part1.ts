import { extractSecondPartValuesFromRaw } from './common';

export function countNumberOfAppearances(rawValues: string[]): number {
  const values = extractSecondPartValuesFromRaw(rawValues);

  const allowedLengthValues = [2, 3, 4, 7];
  const result = values
    .flat()
    .flat()
    .filter((value) => allowedLengthValues.includes(value.length)).length;

  return result;
}
