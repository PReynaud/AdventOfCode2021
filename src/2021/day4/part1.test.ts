import {
  calculateScoreWinningBoard,
  checkBoardIsWinner,
  extractBoards,
  extractDrawnNumbers,
  markValuesFromDrawnNumber,
} from './common';
import { extractAndRun } from './part1';

describe('Day 4 - Part 1', () => {
  const rawValues = [
    '7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1',
    '',
    '22 13 17 11  0',
    ' 8  2 23  4 24',
    '21  9 14 16  7',
    ' 6 10  3 18  5',
    ' 1 12 20 15 19',
    '',
    ' 3 15  0  2 22',
    ' 9 18 13 17  5',
    '19  8  7 25 23',
    '20 11 10 24  4',
    '14 21 16 12  6',
    '',
    '14 21 17 24  4',
    '10 16 15  9 19',
    '18  8 23 26 20',
    '22 11 13  6  5',
    ' 2  0 12  3  7',
  ];

  it('should extract numbers from raw values', () => {
    const result = extractDrawnNumbers(rawValues);
    expect(result).toEqual([
      7, 4, 9, 5, 11, 17, 23, 2, 0, 14, 21, 24, 10, 16, 13, 6, 15, 25, 12, 22,
      18, 20, 8, 19, 3, 26, 1,
    ]);
  });

  it('should extract boards from raw values', () => {
    const result = extractBoards(rawValues);
    expect(result).toHaveLength(3);
    expect(result[0]).toHaveLength(5);
    expect(result[0][0]).toHaveLength(5);
    expect(result[0][0]).toEqual([
      { value: 22, marked: false },
      { value: 13, marked: false },
      { value: 17, marked: false },
      { value: 11, marked: false },
      { value: 0, marked: false },
    ]);
    expect(result[0][1]).toHaveLength(5);
    expect(result[0][1]).toEqual([
      { value: 8, marked: false },
      { value: 2, marked: false },
      { value: 23, marked: false },
      { value: 4, marked: false },
      { value: 24, marked: false },
    ]);
    expect(result[0][4]).toEqual([
      { value: 1, marked: false },
      { value: 12, marked: false },
      { value: 20, marked: false },
      { value: 15, marked: false },
      { value: 19, marked: false },
    ]);
    expect(result[1][0]).toEqual([
      { value: 3, marked: false },
      { value: 15, marked: false },
      { value: 0, marked: false },
      { value: 2, marked: false },
      { value: 22, marked: false },
    ]);

    console.log(result[2]);
  });

  it('should check if a board is winner or not', () => {
    let toCheck = [
      [
        { value: 22, marked: false },
        { value: 13, marked: false },
        { value: 17, marked: false },
        { value: 11, marked: false },
        { value: 0, marked: false },
      ],
      [
        { value: 8, marked: false },
        { value: 2, marked: false },
        { value: 23, marked: false },
        { value: 4, marked: false },
        { value: 24, marked: false },
      ],
      [
        { value: 21, marked: false },
        { value: 9, marked: false },
        { value: 14, marked: false },
        { value: 16, marked: false },
        { value: 7, marked: false },
      ],
      [
        { value: 6, marked: false },
        { value: 10, marked: false },
        { value: 3, marked: false },
        { value: 18, marked: false },
        { value: 5, marked: false },
      ],
      [
        { value: 1, marked: false },
        { value: 12, marked: false },
        { value: 20, marked: false },
        { value: 15, marked: false },
        { value: 19, marked: false },
      ],
    ];
    expect(checkBoardIsWinner(toCheck)).toBe(false);

    toCheck = [
      [
        { value: 22, marked: false },
        { value: 13, marked: false },
        { value: 17, marked: false },
        { value: 11, marked: false },
        { value: 0, marked: false },
      ],
      [
        { value: 8, marked: false },
        { value: 2, marked: false },
        { value: 23, marked: false },
        { value: 4, marked: false },
        { value: 24, marked: false },
      ],
      [
        { value: 21, marked: true },
        { value: 9, marked: true },
        { value: 14, marked: true },
        { value: 16, marked: true },
        { value: 7, marked: false },
      ],
      [
        { value: 6, marked: false },
        { value: 10, marked: false },
        { value: 3, marked: false },
        { value: 18, marked: false },
        { value: 5, marked: false },
      ],
      [
        { value: 1, marked: false },
        { value: 12, marked: false },
        { value: 20, marked: false },
        { value: 15, marked: false },
        { value: 19, marked: false },
      ],
    ];
    expect(checkBoardIsWinner(toCheck)).toBe(false);
    toCheck[2][4].marked = true;
    expect(checkBoardIsWinner(toCheck)).toBe(true);

    toCheck = [
      [
        { value: 22, marked: false },
        { value: 13, marked: false },
        { value: 17, marked: false },
        { value: 11, marked: false },
        { value: 0, marked: false },
      ],
      [
        { value: 8, marked: false },
        { value: 2, marked: true },
        { value: 23, marked: false },
        { value: 4, marked: false },
        { value: 24, marked: false },
      ],
      [
        { value: 21, marked: false },
        { value: 9, marked: true },
        { value: 14, marked: false },
        { value: 16, marked: false },
        { value: 7, marked: false },
      ],
      [
        { value: 6, marked: false },
        { value: 10, marked: true },
        { value: 3, marked: false },
        { value: 18, marked: false },
        { value: 5, marked: false },
      ],
      [
        { value: 1, marked: false },
        { value: 12, marked: true },
        { value: 20, marked: false },
        { value: 15, marked: false },
        { value: 19, marked: false },
      ],
    ];
    expect(checkBoardIsWinner(toCheck)).toBe(false);
    toCheck[0][1].marked = true;
    expect(checkBoardIsWinner(toCheck)).toBe(true);
  });

  it('should mark found values', () => {
    const board = [
      [
        { value: 22, marked: false },
        { value: 13, marked: false },
        { value: 17, marked: false },
        { value: 11, marked: false },
        { value: 0, marked: false },
      ],
      [
        { value: 8, marked: false },
        { value: 2, marked: false },
        { value: 23, marked: false },
        { value: 4, marked: false },
        { value: 24, marked: false },
      ],
      [
        { value: 21, marked: false },
        { value: 9, marked: false },
        { value: 14, marked: false },
        { value: 16, marked: false },
        { value: 7, marked: false },
      ],
      [
        { value: 6, marked: false },
        { value: 10, marked: false },
        { value: 3, marked: false },
        { value: 18, marked: false },
        { value: 5, marked: false },
      ],
      [
        { value: 1, marked: false },
        { value: 12, marked: false },
        { value: 20, marked: false },
        { value: 15, marked: false },
        { value: 19, marked: false },
      ],
    ];
    markValuesFromDrawnNumber(board, 22);
    expect(board[0][0].marked).toBe(true);

    markValuesFromDrawnNumber(board, 0);
    expect(board[0][4].marked).toBe(true);

    markValuesFromDrawnNumber(board, 21);
    markValuesFromDrawnNumber(board, 18);
    markValuesFromDrawnNumber(board, 19);
    expect(board).toMatchSnapshot();
  });

  it('should calculate score of winning board', () => {
    const example = [
      [
        { value: 14, marked: true },
        { value: 21, marked: true },
        { value: 17, marked: true },
        { value: 24, marked: true },
        { value: 4, marked: true },
      ],
      [
        { value: 10, marked: false },
        { value: 16, marked: false },
        { value: 15, marked: false },
        { value: 9, marked: true },
        { value: 19, marked: false },
      ],
      [
        { value: 18, marked: false },
        { value: 8, marked: false },
        { value: 23, marked: true },
        { value: 26, marked: false },
        { value: 20, marked: false },
      ],
      [
        { value: 22, marked: false },
        { value: 11, marked: true },
        { value: 13, marked: false },
        { value: 6, marked: false },
        { value: 5, marked: true },
      ],
      [
        { value: 2, marked: true },
        { value: 0, marked: true },
        { value: 12, marked: false },
        { value: 3, marked: false },
        { value: 7, marked: true },
      ],
    ];
    expect(calculateScoreWinningBoard(example)).toEqual(188);
  });

  it('should validate the example', () => {
    expect(extractAndRun(rawValues)).toEqual(4512);
  });
});
