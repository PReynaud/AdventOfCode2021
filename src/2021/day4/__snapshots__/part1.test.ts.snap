// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day 4 - Part 1 should mark found values 1`] = `
Array [
  Array [
    Object {
      "marked": true,
      "value": 22,
    },
    Object {
      "marked": false,
      "value": 13,
    },
    Object {
      "marked": false,
      "value": 17,
    },
    Object {
      "marked": false,
      "value": 11,
    },
    Object {
      "marked": true,
      "value": 0,
    },
  ],
  Array [
    Object {
      "marked": false,
      "value": 8,
    },
    Object {
      "marked": false,
      "value": 2,
    },
    Object {
      "marked": false,
      "value": 23,
    },
    Object {
      "marked": false,
      "value": 4,
    },
    Object {
      "marked": false,
      "value": 24,
    },
  ],
  Array [
    Object {
      "marked": true,
      "value": 21,
    },
    Object {
      "marked": false,
      "value": 9,
    },
    Object {
      "marked": false,
      "value": 14,
    },
    Object {
      "marked": false,
      "value": 16,
    },
    Object {
      "marked": false,
      "value": 7,
    },
  ],
  Array [
    Object {
      "marked": false,
      "value": 6,
    },
    Object {
      "marked": false,
      "value": 10,
    },
    Object {
      "marked": false,
      "value": 3,
    },
    Object {
      "marked": true,
      "value": 18,
    },
    Object {
      "marked": false,
      "value": 5,
    },
  ],
  Array [
    Object {
      "marked": false,
      "value": 1,
    },
    Object {
      "marked": false,
      "value": 12,
    },
    Object {
      "marked": false,
      "value": 20,
    },
    Object {
      "marked": false,
      "value": 15,
    },
    Object {
      "marked": true,
      "value": 19,
    },
  ],
]
`;
